### PR TITLE
Create Azure Front Door in front of API Management

### PR DESCRIPTION
**Pull request by paolosalvatori**
# Best Practice Checklist

Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment.
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [X] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

* The folder contains the ARM template, parameters file, a README file with a companion article. This ARM template has been used in a PoC with a real customer.

### CODE REVIEW

File: 201-front-door-api-management/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+          }
+        ]
+      },
+      "metadata": {
+        "description": "Virtual Network configuration."
+      }
+    },
+    "workspaceName": {
+      "type": "string",
+      "defaultValue": "FrontDoorApimSampleLogAnalytics",
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.json:33]: does this need to be globally unique?

File: 201-front-door-api-management/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,1048 @@
...
+      "defaultValue": "FrontDoorApimSampleAfd",
+      "minLength": 5,
+      "maxLength": 64,
+      "metadata": {
+        "description": "Name of the Azure Front Door which is globally unique. Min length: 5. Max length: 64"
+      }
+    },
+    "frontDoorEnforceCertificateNameCheck": {
+      "type": "string",
+      "defaultValue": "Disabled",
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.json:110]: Can we enable this by default and/or provide instructions on how to do so?

File: 201-front-door-api-management/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,1048 @@
...
+      "metadata": {
+        "description": "Specifies whether to allow traffic to API Management public endpoints only from Front Door."
+      }
+    }
+  },
+  "variables": {
+    "nsgId": "[resourceId('Microsoft.Network/networkSecurityGroups', parameters('virtualNetwork').subnets[0].nsg)]",
+    "vnetId": "[resourceId('Microsoft.Network/virtualNetworks', parameters('virtualNetwork').name)]",
+    "apimId": "[resourceId('Microsoft.ApiManagement/service', parameters('apimServiceName'))]",
+    "apimHostname": "[concat(toLower(parameters('apimServiceName')), '.azure-api.net')]",
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.json:211]: use reference() to get the endpoint

File: 201-front-door-api-management/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+    "appInsightsId": "[resourceId('Microsoft.Insights/components', parameters('appInsightsName'))]",
+    "apimSubnetId": "[resourceId('Microsoft.Network/virtualNetworks/subnets', parameters('virtualNetwork').name, parameters('virtualNetwork').subnets[0].name)]",
+    "workspaceId": "[resourceId('Microsoft.OperationalInsights/workspaces/', parameters('workspaceName'))]",
+    "loggerId": "[resourceId('Microsoft.ApiManagement/service/loggers', parameters('apimServiceName'), parameters('appInsightsName'))]",
+    "frontDoorId": "[resourceId('Microsoft.Network/frontdoors', parameters('frontDoorName'))]",
+    "frontDoorBackendPoolId": "[concat(variables('frontDoorId'), '/backendPools/', parameters('frontDoorBackendPool').name)]",
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.json:217]: remove concat and use resourceId for all resourceIds

File: 201-front-door-api-management/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,1048 @@
+{
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.json:1]: See the pipeline results for best practice violations and fix what's reported

File: 201-front-door-api-management/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,1048 @@
...
+        }
+      },
+      "resources": [
+        {
+          "apiVersion": "2017-03-01",
+          "type": "products",
+          "name": "custom",
+          "dependsOn": [
+            "[variables('apimId')]",
+            "[concat(variables('apimId'), '/apis/mockApi')]",
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.json:546]: remove concat and use resourceId

File: 201-front-door-api-management/azuredeploy.parameters.json [hunk 7]
```diff
@@ -0,0 +1,97 @@
+{
```
Comment (bmoore-msft) [inline 201-front-door-api-management/azuredeploy.parameters.json:1]: remove the parameters for which you have defaultValues

File: 201-front-door-api-management/README.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
+# Create Azure Front Door in front of Azure API Management #
```
[the remark this answers is not in the document]
Comment (paolosalvatori) [reply 201-front-door-api-management/README.md:1]: I added buttons and badges to the README file

File: 201-front-door-api-management/azuredeploy.parameters.json [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.1",
+  "parameters": {
+    "workspaceName": {
+      "value": "GEN-UNIQUE-20"
+    },
+    "workspaceSku": {
+      "value": "Standard"
+    },
+    "virtualNetwork": {
```
[the remark this answers is not in the document]
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.parameters.json:11]: template and parameters files were already formatted using 2 spaces, I reformatted these files using 4 spaces.

File: 201-front-door-api-management/scripts/deploy.sh [hunk 10]
```diff
@@ -0,0 +1,118 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (paolosalvatori) [reply 201-front-door-api-management/scripts/deploy.sh]: Removed deploy.sh and scripts folder.

File: 201-front-door-api-management/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,1048 @@
...
+    "appInsightsName": {
+      "type": "string",
+      "defaultValue": "FrontDoorApimSampleAppInsights",
+      "metadata": {
+        "description": "The name of your Application Insights resource used by API Management."
+      }
+    },
+    "apimServiceName": {
+      "type": "string",
+      "defaultValue": "FrontDoorApimSampleApim",
```
[the remark this answers is not in the document]
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Removed the defaultValue for apimServiceName parameter.

File: 201-front-door-api-management/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,1048 @@
...
+    "apimCapacity": {
+      "type": "int",
+      "defaultValue": 1,
+      "metadata": {
+        "description": "Number of deployed units of the SKU."
+      }
+    },
+    "frontDoorName": {
+      "type": "string",
+      "defaultValue": "FrontDoorApimSampleAfd",
```
[the remark this answers is not in the document]
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Removed default value for frontDoorName parameter.

File: 201-front-door-api-management/azuredeploy.parameters.json (hunk 7 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.parameters.json:1]: Done

File: 201-front-door-api-management/azuredeploy.json (hunk 1 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Removed default value for workspaceName parameter.

File: 201-front-door-api-management/azuredeploy.json (hunk 2 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Done.

File: 201-front-door-api-management/azuredeploy.json (hunk 6 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Solved. I replaced these resource ids with the following variables:

"mockApiId": "[resourceId('Microsoft.ApiManagement/service/apis', parameters('apimServiceName'), 'mockApi')]",
        "postmanEchoApiId": "[resourceId('Microsoft.ApiManagement/service/apis', parameters('apimServiceName'), 'postmanEchoApi')]",
        "customProductId": "[resourceId('Microsoft.ApiManagement/service/products', parameters('apimServiceName'), 'custom')]",
        "testMethodId": "[resourceId('Microsoft.ApiManagement/service/apis/operations', parameters('apimServiceName'), 'mockApi', 'test')]"

File: 201-front-door-api-management/azuredeploy.json (hunk 4 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Solved. I replaced the previous resource Ids variables with the following:
"frontDoorId": "[resourceId('Microsoft.Network/frontdoors', parameters('frontDoorName'))]",
        "frontDoorBackendPoolId": "[resourceId('Microsoft.Network/frontdoors/backendPools', parameters('frontDoorName'), parameters('frontDoorBackendPool').name)]",
        "frontDoorHealthProbeSettingsId": "[resourceId('Microsoft.Network/frontdoors/healthProbeSettings', parameters('frontDoorName'), parameters('frontDoorHealthProbeSettings').name)]",
        "frontDoorLoadBalancerId": "[resourceId('Microsoft.Network/frontdoors/loadBalancingSettings', parameters('frontDoorName'), parameters('frontDoorBackendPool').loadBalancerName)]",
        "frontDoorFrontedEndpointId": "[resourceId('Microsoft.Network/frontdoors/frontendEndpoints', parameters('frontDoorName'), parameters('frontDoorFrontendEndpoint').name)]",
        "wafPolicyId": "[resourceId('Microsoft.Network/FrontDoorWebApplicationFirewallPolicies', parameters('wafPolicyName'))]",

File: 201-front-door-api-management/azuredeploy.json (hunk 3 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json]: Done. I used the following expression:

"[reference(resourceId('Microsoft.ApiManagement/service', parameters('apimServiceName'))).hostnameConfigurations[0].hostName]"

File: 201-front-door-api-management/azuredeploy.json (hunk 5 above)
Comment (paolosalvatori) [reply 201-front-door-api-management/azuredeploy.json:1]: Done. I also redeployed the template to make sure that it works after all the changes I made.

**Comment by paolosalvatori** [comment]
Hi @bmoore-msft it looks like the deployment fails in USGov. It's Azure Front Door deployment to fail: We couldn't find a DNS record for custom domain ci620af1cb3deb4991b150.azurefd.net that points to Front Door ci620af1cb3deb4991b150.To map a domain to this Front Door, create a CNAME record with your DNS provider for custom domain that points to Front Door." I think this is a problem with Azure Front Door in USGov, but the deployment succeeds in Azure Public. I did all the changes you asked for.

**Comment by bmoore-msft** [comment]
I think there's a "setup" problem - the dns for azurefd is different between the clouds, I made a change to account for that and can get it to fail consistently...  what are the fd dns requirements for this deployment?  Normally you need a CNAME record, and there is 1 for the repo (see contribution) guide, not sure if we need another for Azure Gov...

**Comment by paolosalvatori** [comment]
@bmoore-msft there's no special requirement for this deployment, the ARM template only creates the default frontend host, which is a subdomain of azurefd.net. It does not create any custom domain as explained at https://docs.microsoft.com/en-us/azure/frontdoor/front-door-custom-domain. This would require a custom CNAME, but the default deployment shouldn't. I will ask the Azure Front Door team whether Front Door supports USGov. If you check the other templates for Front Door here in Azure Quickstart Templates, none of them pass the US Gove Test :(

**Comment by bmoore-msft** [comment]
FD wasn't always supported in Gov, so this is new... it looks like it is now but the requirements aren't clear.  Right now, seems to be related to the hostname which is failing both environments.  We can create the CNAME if needed, but need to get the steps to do it figured out.

**Comment by paolosalvatori** [comment]
@bmoore-msft  I fixed the issue and now the ARM template deployed successfully in both Public and US Gov environments. Look at the internal email for more info!

**Pull request merged**